pkg/chunked/cache_linux.go

```diff
@@ -300,7 +300,9 @@ func (c *layersCache) load() error {
 	for _, l := range loadedLayers {
 		l.release()
```

Comment:
AFAICS multiple `load` operations can now be happening in parallel; in that case this might release the same layer multiple times, concurrently.

(I have no opinion on an alternative design at this point, I didn’t look for one yet.)

Reply:
could we just drop the release() here and let the GC deal with it?  We are already using `runtime.SetFinalizer()`

Reply:
That would probably _work_… intuitively I don’t like relying on the GC that way, but I don’t have _much_ of an objective reason (at best, that the time of release, and total memory consumption, are hard to manage).

---

It’s not quite sufficient, though, especially for a single-operation command like `podman pull`. Normally, c/image would trigger a pull of up to 6 layers in parallel, all of those would ~concurrently trigger `layersCache.load`, and at least _read_/map all those cache files into memory, or in the worst case create them from scratch, up to 6 times in parallel.

So before this PR we build/load the cache for every pulled layer; with this PR as is, we build/load the cache for up to 6 concurrent layers (and after the first batch of layers, all future cache uses are only the metadata check, and/or update for the just-added layers). This PR _is_ an improvement but not fully where we want to be.

Ideally, I think, the cache loading / update, at least in the case where we do some I/O, should be serialized.

Would two locks work? Nested “exclusive layer loading lock” → c/storage store+layer locks → “the current `layersCache.mutex”? It seems to me that that should be possible, but I didn’t try and I didn’t check very hard.

---

Very vaguely, the current reference count vs. per-cache mutex vs. loading state seems badly structured, or at least it’s unclear to me why it is so. `getLayersCacheRef` is already serialized and it _knows_ that it needs the cache is old. But then `getLayersCache` triggers a `load` (and a reconciliation against `store.Layers`) _every time_. Do we even need the timer there? Do we need the concurrent existence of _multiple_ `layersCache` objects, with each `chunkedDiffer` keeping a reference to a particular cache version?

I’m not saying it’s _obviously wrong_, and for all I know it might be the best design or the only possible design, but nothing in the code writes down why it is so, and it just … feels … off.